File: Data Structures & Algorithms/course-schedule-ii/submission-2.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> in_degree(numCourses, 0);
        unordered_map<int, vector<int>> adj;
        for (auto &p : prerequisites) {
            int course = p[0];
            int prereq = p[1];
            adj[prereq].push_back(course);
            in_degree[course]++;
        }
        queue<int> bfs;
        for(int i = 0; i < numCourses; ++i){
            if(in_degree[i] == 0){
                bfs.push(i);
            }
        }
        vector<int> res;
        int course_taken = 0;
        while(!bfs.empty()){
            int curr = bfs.front();
            bfs.pop();
            res.push_back(curr);
            course_taken++;
            for(int courses: adj[curr]){
                in_degree[courses]--;
                if(in_degree[courses] == 0){
                    bfs.push(courses);
                }
            }
        }
        if(course_taken == numCourses){
            return res;
        }
        return {};
    }
};

File: Data Structures & Algorithms/course-schedule/submission-2.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        in_degree = [0] * numCourses
        graph = defaultdict(list)
        for course, prereq in prerequisites:
            in_degree[course] += 1
            graph[prereq].append(course)
        
        bfs = deque()
        # visited = [0] * numCourses
        for i in range(0, numCourses):
            if in_degree[i] == 0:
                bfs.append(i)
                # visited[i] = 1

        numCompleted = 0
        while len(bfs) > 0:
            curr = bfs.popleft()
            numCompleted += 1
            for courses in graph[curr]:
                in_degree[courses] -= 1
                if in_degree[courses] == 0:
                    # visited[courses] = 1
                    bfs.append(courses)


        return numCompleted == numCourses